app/src/main/java/com/keylesspalace/tusky/util/SpanUtils.kt

```diff
@@ -0,0 +1,132 @@
+                this.searchCharacter = searchCharacter
+                pattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE)
+                this.searchPrefixWidth = searchPrefixWidth
+            }
+        }
+
+        private fun <T> clearSpans(text: Spannable, spanClass: Class<T>) {
+            for(span in text.getSpans(0, text.length, spanClass)) {
+                text.removeSpan(span)
+            }
+        }
+
+        private fun findPattern(string: String, fromIndex: Int): FindCharsResult {
+            val result = FindCharsResult()
+            for (i in fromIndex..string.lastIndex) {
+                val c = string[i]
+                for (matchType in FoundMatchType.values()) {
+                    val finder = finders[matchType]
+                    if (finder!!.searchCharacter == c
+                            && ((i - fromIndex) < finder.searchPrefixWidth ||
+                                    Character.isWhitespace(string.codePointAt(i - finder.searchPrefixWidth)))) {
+                        result.matchType = matchType
+                        result.start = Math.max(0, i - finder.searchPrefixWidth)
+                        findEndOfPattern(string, result, finder.pattern)
+                        return result
+                    }
+                }
+            }
+            return result
+        }
+
+        private fun findEndOfPattern(string: String, result: FindCharsResult, pattern: Pattern) {
+            val matcher = pattern.matcher(string)
+            if (matcher.find(result.start)) {
+                // Once we have API level 26+, we can use named captures...
+                result.end = matcher.end()
+                result.start = matcher.start()
+                if (Character.isWhitespace(string.codePointAt(result.start))) {
+                    ++result.start
+                }
+            } else {
+                result.end = -1
+            }
+        }
+
+        private fun getSpan(matchType: FoundMatchType, string: String, colour: Int, start: Int, end: Int): CharacterStyle {
+            return when(matchType) {
+                FoundMatchType.HTTP_URL -> CustomURLSpan(string.substring(start, end))
+                FoundMatchType.HTTPS_URL -> CustomURLSpan(string.substring(start, end))
+                else -> ForegroundColorSpan(colour)
+            }
+        }
+
+        /** Takes text containing mentions and hashtags and urls and makes them the given colour. */
+        @JvmStatic
+        fun highlightSpans(text: Spannable, colour: Int) {
+            // Strip all existing colour spans.
+            for (spanClass in spanClasses) {
+                clearSpans(text, spanClass)
+            }
+
+            // Colour the mentions and hashtags.
+            val string = text.toString()
+            val length = text.length
+            var start = 0
+            var end = 0
+            while (end >= 0 && end < length && start >= 0) {
```

Comment:
Studio proposed could of quick-fixes

Reply:
I skipped most of the suggestions (e.g. convert lessthan/greaterthan to range) because this is a hot path and I didn't want to affect the performance too much.

Reply:
Honestly, I didn't check the bytecode but it should be fine. I'm okay with you not changing to ranges